File: solutions/typescript/062-symmetric-tree.ts
/**
 * Problem 62: Symmetric Tree (LeetCode 101)
 * Difficulty: Easy
 * Language: TypeScript
 */
type TreeNode = { val: number; left: TreeNode | null; right: TreeNode | null };

function isSymmetric(root: TreeNode | null): boolean {
  const mirror = (a: TreeNode | null, b: TreeNode | null): boolean => {
    if (a === null && b === null) return true;
    if (a === null || b === null) return false;
    return a.val === b.val &&
      mirror(a.left, b.right) &&
      mirror(a.right, b.left);
  };
  return root === null || mirror(root.left, root.right);
}

File: solutions/rust/062-symmetric-tree.rs
/**
 * Problem 62: Symmetric Tree (LeetCode 101)
 * Difficulty: Easy
 * Language: Rust
 */
fn is_symmetric(root: &Option<Box<TreeNode>>) -> bool {
    fn mirror(a: &Option<Box<TreeNode>>, b: &Option<Box<TreeNode>>) -> bool {
        match (a, b) {
            (None, None) => true,
            (Some(x), Some(y)) =>
                x.val == y.val && mirror(&x.left, &y.right) && mirror(&x.right, &y.left),
            _ => false,
        }
    }
    match root {
        None => true,
        Some(n) => mirror(&n.left, &n.right),
    }
}
